feat($compile): add one-way collection bindings

jbedard opened the pull request:
Implements #14039

jbedard committed:
feat($compile): add one-way collection bindings

Closes #14039

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -6327,6 +6330,54 @@ describe('$compile', function() {
+
+        expect($rootScope.collection[2]).toEqual(newItem);
+      }));
+    });
```
Missing a test to verify the `$watchCollection` part. (AFAICT, these tests would have passed even if `*` was ignored.)

jbedard replied:
I think the first test verifies it, and this test is just wrong/not-applicable (you can tell I copied them both from the `=*` tests...).

gkalpak reviewed (approved):
One minor comment. LGTM otherwise.

petebacondarwin reviewed (approved):
Is it worth testing the `$watchCollection` is being used rather than a deep `$watch`?

jbedard commented:
Updated the tests. Added the one @petebacondarwin suggested, removed the one @gkalpak pointed out was pointless, and added 2 which cause `infdig` with the non-collection version.

jbedard commented:
I wonder if `=*` should be deprecated since `<*` is exactly the same? Right now `=*` has a slightly more complicated action, but I'm pretty sure the logic/perf is the same, and #16550 actually makes it the exact same.

jbedard commented:
So you'd think... but it only does `$watchCollection` on the outer value. So when the inner value changes it won't do anything until the outer value changes, but at that time the outer value should be copied to the inner.

petebacondarwin commented:
No point in deprecating anything now @jbedard  :-)

jbedard commented:
@petebacondarwin I was thinking it would be to encourage using `<*` instead, especially since `=*` is misleading

@gkalpak yeah... unless I'm missing something?

jbedard commented:
Updated the  commit to also close #15874  in addition to #14039

Narretz commented:
There's an eslint error in compile.js (doublequotes)

jbedard commented:
Fixed
